File: src/main/java/com/github/urlshorten/IdStrategy.java
package com.github.urlshorten;

import java.math.BigInteger;
import java.util.function.Function;

public enum IdStrategy implements Function<Long, String> {
    BASE_62 {

        private char[] corpus   = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".toCharArray();

        @Override
        public String apply(Long seed) {
            var number = Long.toString(seed);
            var buf = new char[number.length()];
            var charPos = number.length() - 1;
            var bigIntegerNumber = BigInteger.valueOf(seed);
            var radix = BigInteger.valueOf(corpus.length);

            while (bigIntegerNumber.compareTo(radix) >= 0) {
                buf[charPos--] = corpus[bigIntegerNumber.mod(radix).intValue()];
                bigIntegerNumber = bigIntegerNumber.divide(radix);
            }
            buf[charPos] = corpus[bigIntegerNumber.intValue()];
            return new String(buf, charPos, (number.length() - charPos));
        }
    },
    BASE_10 {
        @Override
        public String apply(Long seed) {
            var value = BigInteger.ZERO;
            for (char c : Long.toString(seed).toCharArray())
            {
                value = value.multiply(BigInteger.valueOf(62));
                if ('0' <= c && c <= '9')
                {
                    value = value.add(BigInteger.valueOf(c - '0'));
                }
                if ('a' <= c && c <= 'z')
                {
                    value = value.add(BigInteger.valueOf(c - 'a' + 10));
                }
                if ('A' <= c && c <= 'Z')
                {
                    value = value.add(BigInteger.valueOf(c - 'A' + 36));
                }
            }
            return value.toString();
        }
    }
}

File: src/main/kotlin/com/github/urlshorten/IdStrategy.kt
package com.github.urlshorten

import java.math.BigInteger
import java.util.function.Function

enum class IdStrategy : Function<Long, String> {
    BASE_62 {
        private val corpus = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".toCharArray()
        override fun apply(seed: Long): String {
            val number = seed.toString()
            val buf = CharArray(number.length)
            var charPos = number.length - 1
            var bigIntegerNumber = BigInteger.valueOf(seed)
            val radix = BigInteger.valueOf(corpus.size.toLong())
            while (bigIntegerNumber >= radix) {
                buf[charPos--] = corpus[bigIntegerNumber.mod(radix).toInt()]
                bigIntegerNumber = bigIntegerNumber.divide(radix)
            }
            buf[charPos] = corpus[bigIntegerNumber.toInt()]
            return String(buf, charPos, number.length - charPos)
        }
    },
    BASE_10 {
        override fun apply(seed: Long): String {
            var value = BigInteger.ZERO
            for (c in java.lang.Long.toString(seed).toCharArray()) {
                value = value.multiply(BigInteger.valueOf(62))
                if (c in '0'..'9') {
                    value = value.add(BigInteger.valueOf((c - '0').toLong()))
                }
                if (c in 'a'..'z') {
                    value = value.add(BigInteger.valueOf((c - 'a' + 10).toLong()))
                }
                if (c in 'A'..'Z') {
                    value = value.add(BigInteger.valueOf((c - 'A' + 36).toLong()))
                }
            }
            return value.toString()
        }
    }
}